### PR TITLE
Add TemporalTimeZone object

### PR DESCRIPTION
**Pull request by gergocs**
Signed-off-by: Gergo Csizi gergocs@inf.u-szeged.hu

### CODE REVIEW

File: src/runtime/TemporalObject.cpp [hunk 1]
```diff
@@ -179,7 +200,7 @@ std::string TemporalObject::tzComponent(ExecutionState& state, std::string& isoS
 
 TemporalObject::DateTime TemporalObject::parseValidIso8601String(ExecutionState& state, std::string isoString, const bool parseTimeZone = false)
 {
-    DateTime dateTime = { 0, 1, 1, 0, 0, 0, 0, 0, 0, "", "", "", "" };
+    DateTime dateTime = { 0, 1, 1, 0, 0, 0, 0, 0, 0, new ASCIIString(), new TimeZone };
```
Comment (clover2123) [inline src/runtime/TemporalObject.cpp:203]: Global variable `String::emptyString` represents an empty string.
So replace allocating a new empty string with this global emptyString value.

File: src/runtime/TemporalObject.cpp [hunk 2]
```diff
@@ -342,8 +335,8 @@ TemporalObject::DateTime TemporalObject::parseValidIso8601String(ExecutionState&
     }
 
     if (parseTimeZone) {
-        if (!dateTime.z.empty()) {
-            dateTime.offsetString = "";
+        if (dateTime.tz->z) {
+            dateTime.tz->offsetString = new ASCIIString("");
```
Comment (clover2123) [inline src/runtime/TemporalObject.cpp:339]: This can be replaced by String::emptyString too.

File: src/runtime/TemporalObject.cpp [hunk 3]
```diff
@@ -513,15 +506,94 @@ TemporalObject::DateTime TemporalObject::parseTemporalInstantString(ExecutionSta
...
+    }
+
+    return result;
+}
+
+TemporalObject::TimeZone* TemporalObject::parseTemporalTimeZoneString(ExecutionState& state, const std::string& isoString)
+{
+    auto result = new TimeZone;
+    result->offsetString = new ASCIIString("");
+    result->name = new ASCIIString("");
```
Comment (clover2123) [inline src/runtime/TemporalObject.cpp:524]: ditto

File: src/runtime/TemporalObject.cpp [hunk 4]
```diff
@@ -513,15 +506,94 @@ TemporalObject::DateTime TemporalObject::parseTemporalInstantString(ExecutionSta
...
+
+    unsigned int index = 0;
+    result = TemporalObject::parseTimeZoneOffset(state, const_cast<std::string&>(isoString), index);
+
+    if (!result || (result->offsetString->length() == 0 && result->name->length() == 0 && !result->z)) {
+        result = TemporalObject::parseValidIso8601String(state, isoString, true).tz;
+    }
+
+    if (result->z) {
+        result->offsetString = new ASCIIString("");
```
Comment (clover2123) [inline src/runtime/TemporalObject.cpp:539]: here too

File: src/runtime/TemporalObject.cpp [hunk 5]
```diff
@@ -513,15 +506,94 @@ TemporalObject::DateTime TemporalObject::parseTemporalInstantString(ExecutionSta
...
+        ErrorObject::throwBuiltinError(state, ErrorObject::RangeError, "bare date-time string is not a time zone");
+    }
 
     return result;
 }
 
+TemporalObject::TimeZone* TemporalObject::parseTimeZoneOffset(ExecutionState& state, std::string& isoString, unsigned int& index)
+{
+    // Offset
+    auto tz = new TemporalObject::TimeZone{ false, new ASCIIString(), new ASCIIString() };
```
Comment (clover2123) [inline src/runtime/TemporalObject.cpp:552]: ditto

**Comment by gergocs** [comment]
Patch is ready for review

File: src/runtime/TemporalObject.cpp [hunk 6]
```diff
@@ -513,15 +504,92 @@ TemporalObject::DateTime TemporalObject::parseTemporalInstantString(ExecutionSta
...
+        ErrorObject::throwBuiltinError(state, ErrorObject::RangeError, "bare date-time string is not a time zone");
+    }
 
     return result;
 }
 
+TemporalObject::TimeZone TemporalObject::parseTimeZoneOffset(ExecutionState& state, std::string& isoString, unsigned int& index)
+{
+    // Offset
+    auto tz = TemporalObject::TimeZone{ false, String::emptyString, String::emptyString };
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:548]: `TimeZone(false, String::emptyString, String::emptyString);` used above. Is this needs to be different?

File: src/runtime/TemporalObject.cpp [hunk 7]
```diff
@@ -1438,6 +1517,221 @@ std::map<TemporalObject::DateTimeUnits, int> TemporalTimeZoneObject::getISOParts
...
+        result += (hours < 10 ? "0" : "") + std::to_string(seconds);
+    }
+
+    return result;
+}
+
+Value TemporalTimeZoneObject::toTemporalTimeZone(ExecutionState& state, const Value& temporalTimeZoneLike)
+{
+    auto timeZoneLike = Value();
+    if (temporalTimeZoneLike.isObject()) {
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:1591]: Newline before

**Review by zherczeg — APPROVED** [review]
LGTM with a few questions

**Comment by gergocs** [comment]
Patch is ready for review

**Review by clover2123 — APPROVED** [review]
LGTM

**Pull request merged**